packages/bindings/src/posts/models.rs

```diff
@@ -0,0 +1,342 @@
+    /// Mention entities.
+    pub mentions: Vec<TagEntity>,
+    /// Url entities.
+    pub urls: Vec<UrlEntity>,
+}
+
+/// Contains the details of a post reference.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct PostReference {
+    /// Type of reference.
+    #[serde(rename = "type")]
+    pub ref_type: PostReferenceType,
+    /// Id of the referenced post.
+    pub post_id: Uint64,
+    /// Position of the reference inside the post's text. This should be used only
+    /// with the type set to ['PostReferenceType::Quote']
+    pub position: Option<Uint64>,
+}
+
+/// Represents the different types of [`PostReference`].
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum PostReferenceType {
+    /// No reference specified.
+    Unspecified,
+    /// Represents a reply to the specified post.
+    ReplayTo,
+    /// Represents a quote of the specified post.
+    Quote,
+    /// Represents a repost of the specified post.
+    Repost,
+}
+
+/// Contains the possible reply settings that a post can have.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum ReplySetting {
+    /// No reply setting specified.
+    #[serde(rename = "REPLY_SETTING_UNSPECIFIED")]
+    Unspecified,
+    /// Everyone will be able to reply to this post.
+    #[serde(rename = "REPLY_SETTING_EVERYONE")]
+    Everyone,
+    /// Only followers of the author will be able to reply to this post.
+    #[serde(rename = "REPLY_SETTING_FOLLOWERS")]
+    Followers,
+    /// Only the author mutual followers will be able to reply to this post.
+    #[serde(rename = "REPLY_SETTING_MUTUAL")]
+    Mutual,
+    /// Only people mentioned inside this post will be able to reply.
+    #[serde(rename = "REPLY_SETTING_MENTIONS")]
+    Mentions,
+}
+
+/// Contains the data of a single post attachment.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct Attachment {
+    /// Id of the subspace inside which the post to which this attachment should be
+    /// connected is.
+    pub subspace_id: Uint64,
+    /// Id of the subspace section inside which the post to which this attachment
+    /// should be connected is.
+    pub section_id: u32,
+    /// Id of the post to which this attachment should be connected.
+    pub post_id: Uint64,
+    /// Id of this attachment.
+    pub id: u32,
+    /// Content of the attachment.
+    pub content: RawPostAttachment,
+}
+
+/// Struct representing a generic post attachment that can be serialized and sent to the chain.  
+/// This struct can be created converting a [`PostAttachment`] using the [`core::convert::Into`] trait.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct RawPostAttachment {
```

Comment:
Can we use enum with `serde(untagged)` here? something like:
```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(untagged)]
pub enum RawPostAttachment{
    Media{
        #[serde(rename = "@type")]
        type_uri: String,
        mime_type: String,
        ...
    }
    Poll{
        #[serde(rename = "@type")]
        type_uri: String,
        question: String,
        ...
    }
}
```
[Reference](https://serde.rs/enum-representations.html#untagged)

Reply:
Yes this can be a solution but I prefer to avoid to expose the `type_uri` field since from a developer point of view can be annoying to type the `Media` and `Poll` type uri.  
The optimal solution should be:
```
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(tag = "@type")]
pub enum RawPostAttachment{
   #[serde(rename = "/desmos.posts.v2.Media")]
    Media{
        mime_type: String,
        ...
    }
   #[serde(rename = "/desmos.posts.v2.Poll")]
    Poll{
        question: String,
        ...
    }
}
```
Unfortunally with this code the VM crash during the deserialization...

I can try to write a custom implementation of the `Serialize` and `Deserialize` traits to overcome this issue

Reply:
@manu0466 
Yeah that is good point, but type uri might be changed to other version like `/desmos.posts.v3.Media` from v2 one. If we make it into a field then the contract must require a upgrade.

Reply:
Yes, but is not garated that the structs doesn't change if the version changes.
What do you think @bragaz?

Reply:
@manu0466 Can't we just have a bunch of constructors method that build the `Media` or `Poll` for the dev and convert them to the `RawContent`?

Reply:
I agree with @bragaz, that we can have a converter which probably is `Into` trait to transform `Media` and `Poll` into `RawContent`. Then we can set type uri in the convert process.

Reply:
@bragaz @dadamu There is already an implementation of `From` to convert a `PostAttachment` into `RawPostAttachment` and also an implmentation of `TryFrom` to convert a `RawPostAttachment` into a `PostAttachment`.
See:
* [From PostAttachment to RawPostAttachment](https://github.com/desmos-labs/desmos-bindings/blob/5a635a6c00ac8961d2f8be210d419a8d1c67d53e/packages/bindings/src/posts/models.rs#L264)
* [TryFrom RawPostAttachment to PostAttachment](https://github.com/desmos-labs/desmos-bindings/blob/5a635a6c00ac8961d2f8be210d419a8d1c67d53e/packages/bindings/src/posts/models.rs#L316)

Reply:
I don't think we have much choice when it comes to change structs if the code changes. Until cosmWasm doesn't fully support protobuf, we'll be a little doomed to change code frequently.

This said, @manu0466 you're right about the traits, devs can use those when building their object. 
Note that we will also write a bit of docs on how to interact and use bindings in our docs website so devs will be aware of these stuff once they've read it.

About the `type_uri` of the `Media` and `Poll`, since they can be subject to changes and go through the converter code each time to change them can be tedious, wouldn't be good to have them as constants? This way you identify them better in the file and it's easier to edit them whenever the type changes.

Reply:
I think that having an enum make the developer experience better, imagin you have to convert a `RawPostAttachment` and do different think if is `Poll` or `Media` the resuling code will be something like this:
```
let raw: RawPostAttachment = ...;
let media = Media::try_from(raw);
let poll = Poll::try_from(raw.clone()); // Extra copy since raw was taked by Media::from

if media.is_ok() {
  let media = media.unwrap();
  // Process the media struct...
} else if poll.is_ok() {
  let poll = poll.unwrap();
  // Process the poll struct...
} else {
  // Error....
}
```

Whit the enum instead will be:
```
let raw: RawPostAttachment = ...;
let post_attachment = PostAttachment::try_from(raw); // No need to copy raw.
if post_attachment.is_ok() {
  match post_attachment.unwrap() {
    case ....
    case ...
  }
} else {
  // Error...
}
```

The situation will become even worse when dealing with a `Vec<PostAttachment>`

Converting the RawPostAttachment into an enum like you did can prevent all this since you already know the type but unluckily the VM crashes during the deserialization

Reply:
I think we should avoid devs write that amount of code (that will become boilerplate each time they need to use poll/media) and give them a solution that allow them to deal with the two types in the easiest way possible at the moment. If for the moment the enum solution doesn't work, we should go for the classic one allowing user to use `type_uri` encapsulated in consts and later when we have more time we can improve the package

Reply:
Thanks for the explanation, I didn't think case that converts from raw attachment. Let's go with enum first.

Reply:
By the way, can we use this [pattern](https://github.com/desmos-labs/desmos-bindings/pull/18/files#r903456502) for the `RawPostAttachment`?

Reply:
With this issue, I found that deserializing works properly on `cosmwasm-1.0.0`. I think we can upgrade it.

Reply:
@dadamu How you tested it? Since when I upgrade to `cosmwasm-1.0.0` I receive an error that tells me that there are some f32 operations in the contract when i try to upload it to the chain

Reply:
@manu0466 I used `cargo unit-test`.